happi/backends/multi_db.py

```diff
@@ -0,0 +1,193 @@
+"""
+Backend implementation that combines multiple backends.
+"""
+import functools
+import logging
+import re
+from typing import Any, Dict, List, Optional, Union
+
+from .core import ItemMeta, ItemMetaGen, _Backend
+
+logger = logging.getLogger(__name__)
+
+
+def prevent_duplicate_ids(fn):
+    """
+    Decorator that remembers which document _id's have been yielded and
+    prevents subsequent documents with matching _id's from being yielded
+
+    Expects the wrapped function to yield ItemMeta documents
+    """
+    @functools.wraps(fn)
+    def wrapper(*args, **kwargs):
+        unique_ids = set()
+        for doc in fn(*args, **kwargs):
+            if doc['_id'] not in unique_ids:
+                unique_ids.add(doc['_id'])
+                yield doc
+
+    return wrapper
+
+
+class MultiBackend(_Backend):
+    """
+    Multi-file backend.
+
+    Combines multiple backends, prioritizing database files in order
+    of appearance.
+
+    Parameters
+    ----------
+    backends : List[_Backend]
+        A list of instantiated backends in order of priority.
+        The backend at index 0 will have the highest priority.
+    """
+    def __init__(self, backends: List[_Backend]):
+        self.backends = backends
+
+    @property
+    def all_items(self) -> List[ItemMeta]:
+        """
+        List of all items in all backends.
+
+        In the case of duplicate items, the item in the first backend
+        takes priority.
+
+        Returns
+        -------
+        List[ItemMeta]
+            A list of all metadata documents
+        """
+        items = []
+        unique_ids = set()
+        for bknd in self.backends:
+            curr_items = bknd.all_items
+            for doc in curr_items:
+                if doc['_id'] not in unique_ids:
+                    items.append(doc)
+                    unique_ids.add(doc['_id'])
+
+        return items
+
+    def clear_cache(self) -> None:
+        """
+        Request to clear any cached data.
+
+        Clears the cache of each backend
+        """
+        for bknd in self.backends:
+            bknd.clear_cache()
+
+    @prevent_duplicate_ids
+    def find(self, *args, multiples=False, **kwargs) -> ItemMetaGen:
+        """
+        Find an instance or instances that matches the search criteria.
+
+        Parameters
+        ----------
+        multiples : bool
+            Find a single result or all results matching the provided
+            information.
+        kwargs
+            Requested information.
+
+        Yields
+        ------
+        ItemMetaGen
+            A generator of metadata documents
+        """
+        for bknd in self.backends:
+            yield from bknd.find(*args, **kwargs)
+
+    def save(self, _id, post, insert=True):
+        """The current implementation of this backend is read-only."""
+        raise NotImplementedError("The Questionnaire backend is read-only")
```

Comment:
Or perhaps the first writable database? Skipping the questionnaire, were it listed first, for example

Reply:
I think it's a bit more complicated than this.

If you save on an existing item, surely it should modify the database source that you loaded that item from?

If you create a new item, the first writable database seems reasonable, but I wonder if this ever gets us into trouble with people accidentally saving their updates in the wrong place. Is it possible that the API needs to be extended in some way to specify which database to include the post in?

Another way to approach this is to say "hey, use a different happi config when you want to make edits" but I wonder if this ends up being difficult.

When used in IPython for example this is pretty reasonable but not for a cli usage. Maybe the cli just needs special handling where it selects which specific backend to write to.

Reply:
I was unsure about this.  If we pick the first writabale / non-duplicate database, I'd be worried the user might not know where the Item went.  (since the backend already obfuscates which backend items are read from)

I'm ok implementing it that way, if we think that's acceptable.  I didn't think it'd be a big deal to make this backend read-only and have people make changes via a single-backend client.

Reply:
I can see the different scenarios for this and agree that it's not quite so clear-cut or obvious how this should be implemented.

Reply:
I vote for no handling in this backend, and you can go call the specific backend you want to save to. The cli can have an extra arg for saving where you specify which config to save to.

Reply:
No-writing for this backend seems good then.  We can tackle saving in the CLI separately, it's reasonable to list out the different backends for the user to pick from.